src/basic/mountpoint-util.c

```diff
@@ -333,31 +333,23 @@ int fd_is_mount_point(int fd, const char *filename, int flags) {
 
         /* Hmm, so, the mount ids are the same. This leaves one special case though for the root file
          * system. For that, let's see if the parent directory has the same inode as we are interested
-         * in. Hence, let's also do fstat() checks now, too, but avoid the st_dev comparisons, since they
-         * aren't that useful on unionfs mounts. */
-        check_st_dev = false;
+         * in. */
+
+        struct stat a, b;
 
-fallback_fstat:
         /* yay for fstatat() taking a different set of flags than the other _at() above */
-        if (flags & AT_SYMLINK_FOLLOW)
-                flags &= ~AT_SYMLINK_FOLLOW;
-        else
-                flags |= AT_SYMLINK_NOFOLLOW;
-        if (fstatat(fd, filename, &a, flags) < 0)
+        if (fstatat(fd, filename, &a, at_flags_mangle_follow_nofollow(flags)) < 0)
                 return -errno;
 
-        if (isempty(filename))
+        if (STR_IN_SET(filename, "", "."))
                 r = fstatat(fd, "..", &b, 0);
         else
                 r = fstatat(fd, "", &b, AT_EMPTY_PATH);
         if (r < 0)
                 return -errno;
 
         /* A directory with same device and inode as its parent? Must be the root directory */
-        if (stat_inode_same(&a, &b))
-                return 1;
-
-        return check_st_dev && (a.st_dev != b.st_dev);
+        return stat_inode_same(&a, &b);
```

Comment:
maybe use inode_same_at() here now?

Reply:
inode_same_at() would internally use name_to_handle_at_try_fid() to compare inodes first, which is redundant here (I'll add a comment)